EIPS/eip-4804.md

```diff
@@ -0,0 +1,163 @@
+
+- **Flexible**.  The contract is able to override the encoding rule so that the contract has fine-control of understanding the actual Web resources that the users want to locate.
+
+### Examples
+
+#### Example 1
+
+```
+web3://ensdomains.eth:4
+```
+
+The protocol will find the address of **ensdomains.eth** from ENS in chainid 4 (Rinkeby), and then call the address with "From" = "0x..." and "Calldata" = "0x" with chainid = 4.
+
+#### Example 2
+
+```
+web3://0x9e081Df45E0D167636DB9C61C7ce719A58d82E3b:4
+```
+
+The protocol will call the address with "From" = "0x9e081Df45E0D167636DB9C61C7ce719A58d82E3b" and "Calldata" = "0x" with chainid = 4.
+
+#### Example 3
+
+```
+web3://cyberbrokers-meta.eth:1/renderBroker/9999
+```
+
+The protocol will find the address of **cyberbrokers-meta.eth** from ENS, and then call the address with "From" = "0x..." and "Calldata" = "0x" + `keccak("view(uint256)")[0:4] + abi.encode(uint256(9999))` with chainid = 1.
+
+#### Exmaple 4
+
+```
+web3://home.w3q:3334/
+```
+
+The protocol will find the address of **home.w3q** from W3NS, and then call the address with "From" = "0x..." and "Calldata" = "0x" with chainid = 3334.
+
+#### Example 5
+
+```
+web3://wusdt.eth:4->(uint256)/balanceOf/charles.eth
+```
+
+The protocol will find the addresses of **wusdt.eth** and **charles.eth** and then call the method "balanceOf(address)" of the contract with the **charles.eth**'s address. The returned data will be parsed as uint256 like `[ "10000000000000" ]`.
+
+#### Example 6
+```
+web3://wusdt.eth:4->()/balanceOf/charles.eth
+```
+
+The protocol will find the address of **wusdt.eth** and then call the method "balanceOf(address)" of the address. The returned data will be parsed as raw bytes like `["0x000000000000000000000000000000000000000000000000000009184e72a000"]`.
+
+### Resolving Address from Name Service Providers
+
+The protocol currently supports two name services:
+- Ethereum name service with suffix `.eth`; and
+- W3 name service with suffix `.w3q`.
+
+#### Resolving Address from ENS
+Given **contractName**, the protocol will find the address of the contract using the following steps:
+1. Find the `w3q` text record on ENS resolver.  Return error if the record is an invalid ETH address.
+2. If the `w3q` text record does not exist, the protocol will use ETH address of the name.
+
+Given **name** in arguments, the protocol will find the address of the name using its ETH address.
+
+#### Resolving Address from W3NS
+
+Given **contractName**, the protocol will find the address of the contract using the following steps:
+1. Find the Web handler record on W3NS resolver.  Return error if the record is an invalid ETH address.
+2. If the Web handler record does not exist, the protocol will use ETH address of the name.
+
+Given **name** in arguments, the protocol will find the address of the name using its ETH address.
```

Comment:
These should be in `Specification` section.

Reply:
Done.  Thanks for the comment.